pkg/controller/miganalytic/restic_executor.go

```diff
@@ -0,0 +1,156 @@
+func (r *ResticDFCommandExecutor) DF(podRef *corev1.Pod, persistentVolumes []MigAnalyticPersistentVolumeDetails) DFCommand {
+	// TODO: use the appropriate block size based on PVCs
+	dfCmd := DFCommand{
+		BaseLocation: "/host_pods",
+		BlockSize:    DecimalSIMega,
+		StdOut:       "",
+		StdErr:       "",
+	}
+	cmdString := dfCmd.PrepareDFCommand(persistentVolumes)
+	restCfg := r.Client.RestConfig()
+	podCommand := pods.PodCommand{
+		Pod:     podRef,
+		RestCfg: restCfg,
+		Args:    cmdString,
+	}
+	err := podCommand.Run()
+	if err != nil {
+		log.Info(
+			fmt.Sprintf("df command inside pod %s returned non-zero error code", podRef.Name))
+	}
+	dfCmd.StdErr = podCommand.Err.String()
+	dfCmd.StdOut = podCommand.Out.String()
+	return dfCmd
+}
+
+// getResticPodForNode lookup Restic Pod ref in local cache
+func (r *ResticDFCommandExecutor) getResticPodForNode(nodeName string) *corev1.Pod {
+	if podRef, exists := r.ResticPodReferences[nodeName]; exists {
+		return podRef
+	}
+	return nil
+}
+
+// loadResticPodReferences load Restic Pod refs in-memory
+func (r *ResticDFCommandExecutor) loadResticPodReferences() error {
+	if r.ResticPodReferences == nil {
+		r.ResticPodReferences = make(map[string]*corev1.Pod)
+	}
+	resticPodList := corev1.PodList{}
+	labelSelector := client.InNamespace(r.Namespace).MatchingLabels(
+		map[string]string{
+			ResticPodLabelKey: ResticPodLabelValue})
+	err := r.Client.List(context.TODO(), labelSelector, &resticPodList)
+	if err != nil {
+		return err
+	}
+	for i := range resticPodList.Items {
+		if resticPodList.Items[i].Spec.NodeName != "" {
+			r.ResticPodReferences[resticPodList.Items[i].Spec.NodeName] = &resticPodList.Items[i]
+		}
+	}
+	return nil
+}
+
+// Execute given a map node->[]pvc, runs Df command for each, returns list of structured df output per pvc
+func (r *ResticDFCommandExecutor) Execute(pvcNodeMap map[string][]MigAnalyticPersistentVolumeDetails) ([]DFOutput, error) {
+	gatheredData := []DFOutput{}
+	err := r.loadResticPodReferences()
+	if err != nil {
+		return gatheredData, err
+	}
+	// dfOutputs for n nodes
+	dfOutputs := make(map[string]DFCommand, len(pvcNodeMap))
+	waitGroup := sync.WaitGroup{}
+	mutex := sync.Mutex{}
+	// allows setting a limit on number of concurrent df threads running
+	bufferedExecutionChannel := make(chan struct{}, 10)
+	// run df concurrently for 'n' nodes
+	for node := range pvcNodeMap {
+		resticPodRef := r.getResticPodForNode(node)
+		// if no Restic pod is found for this node, all PVCs on this node are skipped
+		if resticPodRef == nil {
+			for _, pvc := range pvcNodeMap[node] {
+				dfOutput := DFOutput{
+					IsError:   true,
+					Name:      pvc.Name,
+					Namespace: pvc.Namespace,
+				}
+				gatheredData = append(gatheredData, dfOutput)
+			}
+			continue
+		}
+		waitGroup.Add(1)
+		go func(n string, podRef *corev1.Pod) {
+			// block until channel empty
+			bufferedExecutionChannel <- struct{}{}
```

Comment:
Worth noting for any other reviewers looking at this scratching their heads, this is just stuffing a blank struct in the channel as a means of messaging. No actual data is contained.